File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        int max_rectangle = 0;
        stack<pair<int,int>> stk;
        for(int i = 0; i < heights.size(); i++)
        {
            int index = i;
            while(!stk.empty() and stk.top().second > heights[i])
            {
                pair<int,int> p = stk.top();
                index = p.first;
                max_rectangle = max(max_rectangle, p.second * (i - p.first));
                stk.pop();
            }
            stk.push({index,heights[i]});
        }
        while(!stk.empty())
        {
            pair<int,int> p = stk.top();
            max_rectangle = max(max_rectangle, p.second * (static_cast<int>(heights.size()) - p.first));
            stk.pop();
        }
        return max_rectangle;
    }
};

File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.cs
public class Solution {
    public int LargestRectangleArea(int[] heights)
    {
        var stack = new Stack<(int indx, int heigth)>();
        int max_area = 0;
        for (int i = 0; i < heights.Length; i++)
        {
            if( stack.Count() > 0)
            {
                int top_height = stack.Peek().heigth;
                int top_indx = stack.Peek().indx;
                int min_indx = -1;
                while (stack.Count() > 0 && top_height > heights[i])
                {
                    max_area = Math.Max(max_area, (i - top_indx) * top_height);
                    stack.Pop();
                    min_indx = top_indx;

                    stack.TryPeek(out (int indx,int heigth) pair);
                    top_height = pair.heigth;
                    top_indx = pair.indx;
                }
                if (min_indx != -1)
                    stack.Push((min_indx, heights[i]));
                else
                    stack.Push((i, heights[i]));
            }
            else
            {
                stack.Push((i, heights[i]));
            }
        }
        while (stack.Count() > 0)
        {
            int top_height = stack.Peek().heigth;
            int top_indx = stack.Peek().indx;
            max_area = Math.Max(max_area, (heights.Length - top_indx) * top_height);
            stack.Pop();
        }
        return max_area;
    }
}
